bindings/java/java/src/main/java/org/ethereum/evmc/Host.java

```diff
@@ -70,16 +70,17 @@ static ByteBuffer get_code_hash(int context_index, byte[] address) {
   }
 
   /** Copy code callback function. */
-  static ByteBuffer copy_code(int context_index, byte[] address, int code_offset) {
+  static ByteBuffer copy_code(int context_index, byte[] address) {
     HostContext context =
         requireNonNull(
             getContext(context_index),
             "HostContext does not exist for context_index: " + context_index);
     byte[] code = context.getCode(address).array();
 
-    if (code != null && code_offset > 0 && code_offset < code.length) {
-      int length = code.length - code_offset;
-      return ByteBuffer.allocateDirect(length).put(code, code_offset, length);
+    /* Ensure code is returned and both offset/length are positive numbers. */
+    if (code != null) {
+      // TODO: use wrap here, no need to duplicate this given it is copied again.
```

Comment:
This TODO says we can use ByteBuffer.wrap, am I reading this right? But for C code to interact with ByteBuffer, you need to use a direct memory allocation. I had panics when I tried to use heap-based bytebuffers.

Reply:
I did not see this difference in the JNI docs, only whether it is copied or not. And here we do not need copying given the C counterpart does a copy anyway.

Can you point me to the documentation where this is mentioned?

Reply:
I think the code used to get the pointer of the bytebuffer, and that would crash the JVM. So I think you're saying that's all changed now - I'm good with a change there, and I'll make sure to give you a reproduction case if that happens again.

Reply:
I think you are right, documentation does seem to claim that `allocateDirect` will 100% be laid on a way it can be accessed without copying, while other may require copying. It is not clear however who does that copying, but it seems it is done by the JVM so it would mean a slowdown.